Objects/moduleobject.c

```diff
@@ -784,11 +786,73 @@ _PyModuleSpec_IsUninitializedSubmodule(PyObject *spec, PyObject *name)
     return rc;
 }
 
+static int
+_is_module_possibly_shadowing(PyObject *origin)
+{
+    // origin must be a unicode subtype
+    // Returns 1 if the module at origin could be shadowing a module of the
+    // same name later in the module search path. The condition we check is basically:
+    // not sys.flags.safe_path and os.path.dirname(origin) == (sys.path[0] or os.getcwd())
+    // Returns 0 otherwise (or if we aren't sure)
+    // Returns -1 if an error occurred that should be propagated
+    if (origin == NULL) {
+        return 0;
+    }
+    const PyConfig *config = _Py_GetConfig();
+    if (config->safe_path) {
+        return 0;
+    }
+
+    // os.path.dirname(origin)
+    wchar_t origin_dirname[MAXPATHLEN + 1];
+    Py_ssize_t size = PyUnicode_AsWideChar(origin, origin_dirname, MAXPATHLEN);
+    if (size < 0) {
+        return -1;
+    }
+    assert(size <= MAXPATHLEN);
+    origin_dirname[size] = L'\0';
+
+    wchar_t *sep = wcsrchr(origin_dirname, SEP);
+    if (!sep) {
+        return 0;
+    }
+    *sep = L'\0';
+
+    // sys.path[0] or os.getcwd()
+    wchar_t *sys_path_0 = config->sys_path_0;
+    if (!sys_path_0) {
+        if (!config->module_search_paths_set || config->module_search_paths.length == 0) {
+            return 0;
+        }
+        sys_path_0 = config->module_search_paths.items[0];
+        assert(sys_path_0 != NULL);
```

Comment:
I was wrong about this.  `config->sys_path_0` is strictly distinct from `config->module_search_paths.items[0]`.  So we shouldn't use it as a fallback.

```suggestion
        return 0;
```